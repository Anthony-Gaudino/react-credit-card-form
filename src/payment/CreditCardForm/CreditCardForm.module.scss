@use "../../theme";

.creditCardForm {
  max-width: 570px;

  &__form {
    display: flex;
    flex-direction: column;
    gap: 1rem;
    padding: 10rem 1.5rem 1.5rem;
    margin-top: -8rem;
  }

  &__inline {
    display: flex;
    gap: 1rem;
    align-items: flex-start;

    &--large {
      gap: 2rem;
    }
  }

  &__expiration-date-wrapper {
    flex: 2;
  }

  &__cvv-wrapper {
    flex: 1;
  }

  &__submit-button-wrapper {
    margin-top: 1rem;
  }
}
