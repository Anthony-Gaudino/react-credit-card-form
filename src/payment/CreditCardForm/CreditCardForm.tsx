import { useState } from "react";

import Button from "../../components/Button";
import Fieldset from "../../components/Fieldset";
import Paper from "../../components/Paper";
import Select from "../../components/Select";
import TextField from "../../components/TextField";
import CreditCard from "../CreditCard";
import styles from "./CreditCardForm.module.scss";

const CURRENT_YEAR = new Date().getFullYear();

/**
 * Builds a list of select options.
 *
 * @param n - The option index number.
 * @param val - The option value.
 * @param defaultText - The default select text (first item).
 *
 * @returns A select option.
 */
const buildList = (n: number, val: string, defaultText: string) => (
  <option key={n} value={n === 0 ? "" : val} disabled={n === 0}>
    {n === 0 ? defaultText : val}
  </option>
);

/**
 * Month picker options (1 to 12).
 */
const monthOptions = [...Array(13).keys()].map((n) =>
  buildList(n, n.toString().padStart(2, "0"), "Month")
);

/**
 * Year picker options, from current year to current +10 years.
 */
const yearOptions = (() =>
  [...Array(11).keys()].map((n) =>
    buildList(n, (CURRENT_YEAR + n - 1).toString(), "Year")
  ))();

/**
 * A form to enter credit card details.
 */
export const CreditCardForm = (): JSX.Element => {
  const [name, setName] = useState("");
  const [cardNumber, setCardNumber] = useState("");
  const [cardNumberMasked, setCardNumberMasked] = useState("");
  const [cvv, setCvv] = useState("");
  const [selectedMonth, setSelectedMonth] = useState("");
  const [selectedYear, setSelectedYear] = useState("");
  const [triedSubmmitForm, setTriedSubmmitForm] = useState(false);
  const [showBack, setShowBack] = useState(false);

  const onNameChangeHandler = (e: React.FormEvent<HTMLInputElement>) => {
    setName(e.currentTarget.value);
  };

  const onCardNumberChangeHandler = (e: React.FormEvent<HTMLInputElement>) => {
    const plainNumber = e.currentTarget.value.replace(/[^0-9]/g, "");

    setCardNumber(plainNumber);
    setCardNumberMasked(plainNumber.replace(/\d{4}(?=.)/g, "$& "));
  };

  const onCvvChangeHandler = (e: React.FormEvent<HTMLInputElement>) => {
    setCvv(e.currentTarget.value.replace(/[^0-9]/g, ""));
  };

  const onMonthChangeHandler = (e: React.FormEvent<HTMLSelectElement>) => {
    setSelectedMonth(e.currentTarget.value);
  };

  const onYearChangeHandler = (e: React.FormEvent<HTMLSelectElement>) => {
    setSelectedYear(e.currentTarget.value);
  };

  const onFrontCardFieldFocus = () => setShowBack(false);

  const onBackCardFieldFocus = () => setShowBack(true);

  const onCardClickHandler = () => setShowBack(!showBack);

  const onSubmitButtonClickHandler = (
    e: React.MouseEvent<HTMLButtonElement>
  ) => {
    if (!triedSubmmitForm) setTriedSubmmitForm(true);
  };

  const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const values = {
      name,
      cardNumber,
      cvv,
      month: selectedMonth,
      year: selectedYear,
    };

    // Do something with `values`...
  };

  return (
    <div className={styles.creditCardForm}>
      <CreditCard
        name={name}
        cardNumber={cardNumber}
        printedCardNumber={cardNumberMasked}
        month={selectedMonth}
        year={selectedYear}
        cvv={cvv}
        showBack={showBack}
        cardClickHandler={onCardClickHandler}
      />
      <Paper elevation={6}>
        <form className={styles.creditCardForm__form} onSubmit={handleSubmit}>
          <TextField
            type="text"
            label="Card Number"
            value={cardNumberMasked}
            required
            minLength={16}
            maxLength={23}
            inputMode="numeric"
            pattern="[0-9 ]+"
            highlightInvalid={triedSubmmitForm}
            onChange={onCardNumberChangeHandler}
            onFocus={onFrontCardFieldFocus}
          />
          <TextField
            type="text"
            label="Card Name"
            value={name}
            required
            minLength={3}
            maxLength={22}
            highlightInvalid={triedSubmmitForm}
            onChange={onNameChangeHandler}
            onFocus={onFrontCardFieldFocus}
          />
          <div
            className={`${styles.creditCardForm__inline} ${styles["creditCardForm__inline--large"]}`}
          >
            <div className={styles["creditCardForm__expiration-date-wrapper"]}>
              <Fieldset legend="Expiration Date">
                <div className={styles.creditCardForm__inline}>
                  <Select
                    label="Month"
                    hideLabel={true}
                    value={selectedMonth}
                    required
                    highlightInvalid={triedSubmmitForm}
                    onChange={onMonthChangeHandler}
                    onFocus={onFrontCardFieldFocus}
                  >
                    {monthOptions}
                  </Select>
                  <Select
                    label="Year"
                    hideLabel={true}
                    value={selectedYear}
                    required
                    highlightInvalid={triedSubmmitForm}
                    onChange={onYearChangeHandler}
                    onFocus={onFrontCardFieldFocus}
                  >
                    {yearOptions}
                  </Select>
                </div>
              </Fieldset>
            </div>
            <div className={styles["creditCardForm__cvv-wrapper"]}>
              <TextField
                type="text"
                label="CVV"
                value={cvv}
                required
                minLength={3}
                maxLength={4}
                inputMode="numeric"
                pattern="[0-9]+"
                highlightInvalid={triedSubmmitForm}
                onChange={onCvvChangeHandler}
                onFocus={onBackCardFieldFocus}
              />
            </div>
          </div>
          <div className={styles["creditCardForm__submit-button-wrapper"]}>
            <Button fullWidth onClick={onSubmitButtonClickHandler}>
              Submit
            </Button>
          </div>
        </form>
      </Paper>
    </div>
  );
};
