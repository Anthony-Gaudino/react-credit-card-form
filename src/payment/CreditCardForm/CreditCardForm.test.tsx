import { render, screen } from "@testing-library/react";

import CreditCardForm from ".";

describe("Test CreditCardForm component", () => {
  test("the form is correctly rendered", () => {
    const { asFragment } = render(<CreditCardForm />);

    const nameField = screen.getByLabelText("Card Name");
    const cardNumberField = screen.getByLabelText("Card Number");
    const monthField = screen.getByLabelText("Month");
    const yearField = screen.getByLabelText("Year");
    const cvvField = screen.getByLabelText("CVV");

    expect(nameField).toBeInTheDocument();
    expect(cardNumberField).toBeInTheDocument();
    expect(monthField).toBeInTheDocument();
    expect(yearField).toBeInTheDocument();
    expect(cvvField).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
