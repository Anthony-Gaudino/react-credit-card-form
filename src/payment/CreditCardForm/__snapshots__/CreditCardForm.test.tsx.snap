// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CreditCardForm component the form is correctly rendered 1`] = `
<DocumentFragment>
  <div
    class="creditCardForm"
  >
    <div
      class="creditCard"
      role="button"
      tabindex="-1"
    >
      <div
        class="paper creditCard__paper"
        style="--elevation: 6;"
      >
        <div
          class="creditCard__surface creditCard__surface--front"
        >
          <div
            class="creditCard__contentWrapper"
          >
            <div
              class="creditCard__row"
            >
              <img
                alt=""
                class="creditCard__chip"
                src="smart_chip.svg"
              />
              <img
                alt=""
                class="creditCard__issuer"
                src="[object Object]"
              />
            </div>
            <div
              class="creditCard__row"
            >
              <span
                class="creditCard__number"
              />
            </div>
            <div
              class="creditCard__row"
            >
              <div
                class="creditCard__info"
              >
                <div
                  class="creditCard__infoTitle"
                >
                  Card Holder
                </div>
                <div
                  class="creditCard__infoContent"
                />
              </div>
              <div
                class="creditCard__info"
              >
                <div
                  class="creditCard__infoTitle"
                >
                  Expires
                </div>
                <div
                  class="creditCard__infoContent"
                >
                  MM/YY
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="creditCard__surface creditCard__surface--back"
        >
          <div
            class="creditCard__contentWrapper"
          >
            <div
              class="creditCard__magneticTape"
            />
            <div
              class="creditCard__signatureAndCVVWrapper"
            >
              <div
                class="creditCard__CVVTitle"
              >
                CVV
              </div>
              <div
                class="creditCard__CVVField"
              >
                <span />
              </div>
            </div>
            <div
              class="creditCard__row creditCard__row--right"
            >
              <img
                alt=""
                class="creditCard__issuer"
                src="[object Object]"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="paper undefined"
      style="--elevation: 6;"
    >
      <form
        class="creditCardForm__form"
      >
        <label
          class="inputLabel"
        >
          Card Number
          <input
            class="textField"
            inputmode="numeric"
            maxlength="23"
            minlength="16"
            pattern="[0-9 ]+"
            required=""
            type="text"
            value=""
          />
        </label>
        <label
          class="inputLabel"
        >
          Card Name
          <input
            class="textField"
            maxlength="22"
            minlength="3"
            required=""
            type="text"
            value=""
          />
        </label>
        <div
          class="creditCardForm__inline creditCardForm__inline--large"
        >
          <div
            class="creditCardForm__expiration-date-wrapper"
          >
            <fieldset
              class="fieldset"
            >
              <legend
                class="fieldset__legend"
              >
                Expiration Date
              </legend>
              <div
                class="creditCardForm__inline"
              >
                <label
                  class="inputLabel inputLabel--hidden"
                >
                  Month
                  <select
                    class="select"
                    required=""
                  >
                    <option
                      disabled=""
                      value=""
                    >
                      Month
                    </option>
                    <option
                      value="01"
                    >
                      01
                    </option>
                    <option
                      value="02"
                    >
                      02
                    </option>
                    <option
                      value="03"
                    >
                      03
                    </option>
                    <option
                      value="04"
                    >
                      04
                    </option>
                    <option
                      value="05"
                    >
                      05
                    </option>
                    <option
                      value="06"
                    >
                      06
                    </option>
                    <option
                      value="07"
                    >
                      07
                    </option>
                    <option
                      value="08"
                    >
                      08
                    </option>
                    <option
                      value="09"
                    >
                      09
                    </option>
                    <option
                      value="10"
                    >
                      10
                    </option>
                    <option
                      value="11"
                    >
                      11
                    </option>
                    <option
                      value="12"
                    >
                      12
                    </option>
                  </select>
                </label>
                <label
                  class="inputLabel inputLabel--hidden"
                >
                  Year
                  <select
                    class="select"
                    required=""
                  >
                    <option
                      disabled=""
                      value=""
                    >
                      Year
                    </option>
                    <option
                      value="2022"
                    >
                      2022
                    </option>
                    <option
                      value="2023"
                    >
                      2023
                    </option>
                    <option
                      value="2024"
                    >
                      2024
                    </option>
                    <option
                      value="2025"
                    >
                      2025
                    </option>
                    <option
                      value="2026"
                    >
                      2026
                    </option>
                    <option
                      value="2027"
                    >
                      2027
                    </option>
                    <option
                      value="2028"
                    >
                      2028
                    </option>
                    <option
                      value="2029"
                    >
                      2029
                    </option>
                    <option
                      value="2030"
                    >
                      2030
                    </option>
                    <option
                      value="2031"
                    >
                      2031
                    </option>
                  </select>
                </label>
              </div>
            </fieldset>
          </div>
          <div
            class="creditCardForm__cvv-wrapper"
          >
            <label
              class="inputLabel"
            >
              CVV
              <input
                class="textField"
                inputmode="numeric"
                maxlength="4"
                minlength="3"
                pattern="[0-9]+"
                required=""
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          class="creditCardForm__submit-button-wrapper"
        >
          <button
            class="button button--full-width"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
